cpp/src/io/orc/dict_enc.cu

```diff
@@ -88,18 +92,13 @@ static __device__ void LoadNonNullIndices(volatile dictinit_state_s *s, int t)
     }
     __syncthreads();
     is_valid = (i + t < s->chunk.num_rows) ? (s->scratch_red[t >> 5] >> (t & 0x1f)) & 1 : 0;
-    nz_map   = ballot(is_valid);
-    nz_pos   = s->nnz + __popc(nz_map & (0x7fffffffu >> (0x1fu - ((uint32_t)t & 0x1f))));
-    if (!(t & 0x1f)) { s->scratch_red[16 + (t >> 5)] = __popc(nz_map); }
+    uint32_t tmp_nnz;
+    cub::BlockScan<uint32_t, block_size, cub::BLOCK_SCAN_WARP_SCANS>(temp_storage)
+      .ExclusiveSum(is_valid, nz_pos, tmp_nnz);
+    nz_pos += s->nnz;
     __syncthreads();
-    if (t < 32) {
-      uint32_t nnz     = s->scratch_red[16 + (t & 0xf)];
-      uint32_t nnz_pos = WarpReducePos16(nnz, t);
-      if (t == 0xf) { s->nnz += nnz_pos; }
-      if (t <= 0xf) { s->scratch_red[t] = nnz_pos - nnz; }
-    }
-    __syncthreads();
-    if (is_valid) { s->dict[nz_pos + s->scratch_red[t >> 5]] = i + t; }
+    if (!t) { s->nnz += tmp_nnz; }
+    if (is_valid) { s->dict[nz_pos] = i + t; }
```

Comment:
Yes,

Here this is an ExclusiveSum in a Warp with 32 threads + `s->nnz`.
```
nz_map   = ballot(is_valid);
nz_pos   = s->nnz + __popc(nz_map & (0x7fffffffu >> (0x1fu - ((uint32_t)t & 0x1f))));
```

This will get sum of valid bits, which is sum of whole warp.
```
if (!(t & 0x1f)) { s->scratch_red[16 + (t >> 5)] = __popc(nz_map); }
```

This part would warp the 16 values got from prev line and create offsets
[10, 10, 10 ....] -> [10, 20, 30, 40]
```
if (t < 32) {
  uint32_t nnz     = s->scratch_red[16 + (t & 0xf)];
  uint32_t nnz_pos = WarpReducePos16(nnz, t);
```
At 0xf, you would be having sum of all warps, which is nothing but sum of all threads of `is_valid`.
Which is being updated to shared object `s->nnz`
```
  if (t == 0xf) { s->nnz += nnz_pos; }
```

Here, the offsets are being updated so that this can be added to the warped values to create block wide sum
[10, 20, 30, 40, ... ] -> [0, 10, 20, 30, ..], 
```
  if (t <= 0xf) { s->scratch_red[t] = nnz_pos - nnz; }
}
```
  
The offsets calculated in prev step are added to get block wide Exclusive sum for each thread.
```
if (is_valid) { s->dict[nz_pos + s->scratch_red[t >> 5]] = i + t; }
```

So, with following code, I would get ExclusiveSum across whole block along with aggregate sum in tmp_nnz
```
cub::BlockScan<uint32_t, block_size, cub::BLOCK_SCAN_WARP_SCANS>(temp_storage).ExclusiveSum(is_valid, nz_pos, tmp_nnz);
```

Update `nz_pos` with current `s->nnz` as they are doing in the warp sum
```
nz_pos += s->nnz;
__syncthreads();
if (!t) { s->nnz += tmp_nnz; }
```

Since `nz_pos` is already block-wide ExclusiveSum, don't have to add anything
```
if (is_valid) { s->dict[nz_pos] = i + t; }
```